src/Microsoft.AspNetCore.Razor.Language/StreamSourceDocument.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal class StreamSourceDocument : RazorSourceDocument
+    {
+        // Internal for testing
+        internal readonly RazorSourceDocument _innerSourceDocument;
+
+        private readonly byte[] _checksum;
+
+        public StreamSourceDocument(Stream stream, Encoding encoding, string fileName)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _checksum = ComputeChecksum(stream);
```

Comment:
So we don't have to hang onto the underlying bytes of the stream. There's a lot of them 😉